core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala

```diff
@@ -130,12 +130,13 @@ private[spark] class PythonWorkerFactory(
         // Pull from idle workers until we one that is alive, otherwise create a new one.
         while (idleWorkers.nonEmpty) {
           val worker = idleWorkers.dequeue()
-          val workerHandle = daemonWorkers(worker)
-          if (workerHandle.isAlive()) {
-            try {
-              return (worker.refresh(), Some(workerHandle))
-            } catch {
-              case c: CancelledKeyException => /* pass */
+          daemonWorkers.get(worker).foreach { workerHandle =>
+            if (workerHandle.isAlive()) {
+              try {
+                return (worker.refresh(), Some(workerHandle))
+              } catch {
+                case _: CancelledKeyException => /* pass */
+              }
             }
           }
```

Comment:
IIUC, after this PR, instead of showing `key not found`, it shows `Worker xxx process from idle queue is dead, discarding.` and stopped the worker directly?

Reply:
Yes, exactly. It will write the warning and stop the out-of-sync worker.

Reply:
Also updated the description. Thanks.